### PR TITLE
Fix initialization of gpuVertexFinder::Producer to avoid crash when cudaMalloc has not been called

### PR DESCRIPTION
**Pull request by makortel**
While rebasing #100 I got stuck for a while debugging crashes/failing asserts. One failure on the course (but not the real cause) was that if `gpuVertexFinder::Producer` gets destructed without its `allocateOnGPU()` being called, `cudaFree()` returns an error and `cudaCheck()` aborts.

A possible fix is to initialize all the device-memory-pointers to `nullptr`. Note that the initialization can not be done in constructor because `cuda::memory::device::make_unique()` requires trivial constructor (as do our `CUDAService::make_device_unique`).

### CODE REVIEW

**Comment by makortel** [comment]
Unfortunately no, any user-defined constructor (even `T() = default`) leads to `std::is_trivially_constructible<T>` to give `false`.

**Comment by fwyzard** [comment]
Mhm, that's not what I see.

**test.cc**
```c++
#include <iostream>
#include <type_traits>

struct A
{
};

struct B
{
  B() noexcept { }
};

struct C
{
  C() = default;
};

struct D
{
  D() = default;
  D(nullptr_t) { }
};

int main() {
  std::cout << std::boolalpha;
  std::cout << "is_trivially_constructible:" << std::endl;
  std::cout << "int: " << std::is_trivially_constructible<int>::value << std::endl;
  std::cout << "A:   " << std::is_trivially_constructible<A>::value << std::endl;
  std::cout << "B:   " << std::is_trivially_constructible<B>::value << std::endl;
  std::cout << "C:   " << std::is_trivially_constructible<C>::value << std::endl;
  std::cout << "D:   " << std::is_trivially_constructible<D>::value << std::endl;
  return 0;
}
```

```bash
$ g++-7 -std=c++11 -O2 test.cc -o test
$ ./test 
is_trivially_constructible:
int: true
A:   true
B:   false
C:   true
D:   true
```

As long as the default constructor is defaulted, it seems that we can have an additional non-default constructor (see also [this definition](https://en.cppreference.com/w/cpp/types/is_trivial) of *TrivialType* and `is_trivial`).

**Comment by makortel** [comment]
@fwyzard Right, I had the mistake of having the defaulted constructor private... I'll add a non-default constructor to do the setting. Thanks for pointing it out!

File: RecoPixelVertexing/PixelVertexFinding/src/gpuVertexFinder.h [hunk 1]
```diff
@@ -66,6 +74,7 @@ namespace gpuVertexFinder {
 	     float ichi2max,   // max normalized distance to cluster
              bool ienableTransfer
 	     ) :
+      onGPU{nullptr},
```
Comment (fwyzard) [inline RecoPixelVertexing/PixelVertexFinding/src/gpuVertexFinder.h]: Would you mind if we use `onGPU(nullptr)` for consistency with the rest ?

I don't have a strong opinion though, I'll merge it like this if you prefer.

**Comment by makortel** [comment]
There you go.

**Comment by fwyzard** [comment]
:construction: Validation running at workergpu09:/mnt/home/fwyzard/validation/run.C0gw720HM0 ...

**Comment by fwyzard** [comment]
No changes in physics, as expected.

**Comment by fwyzard** [comment]
No changes in performance: `1868.2 ± 3.8 ev/s` (before) vs `1868.1 ± 9.7 ev/s` (after).

**Pull request merged**